File: src/modules/users/usecases/create_user_usecase.py

from src.modules.users.ports.output import IUserRepoDep
from src.modules.shared.domain.entities import UserEntity
from src.modules.users.presenter.dtos import (
    CreateUserRequestDto,
    CreateUserResponseDto
)
from src.libs.decorators.transactional_decorator import Transactional
from typing import Annotated
from fastapi import Depends


class CreateUserUsecase():
    def __init__(self, user_repo: IUserRepoDep):
        self.user_repo = user_repo
    
    @Transactional
    async def execute(self, user_dto: CreateUserRequestDto) -> CreateUserResponseDto:
        user_entity = self.user_repo.create(user_dto.to_entity())
        return self.__to_response(user_entity)

    def __to_response(self, user_entity: UserEntity) -> CreateUserResponseDto:
        return CreateUserResponseDto(id=str(user_entity.id))

CreateUserUsecaseDep = Annotated[CreateUserUsecase, Depends(CreateUserUsecase)]

File: src/modules/users/usecases/create-user.usecase.ts
import { Inject, Injectable } from '@nestjs/common';
import { Transactional } from '../../../libs/decorators/transactional.decorator';
import { IUserRepository } from '../ports/output/user-repository.port';
import {
  CreateUserRequestDto,
  CreateUserResponseDto,
} from '../presenter/dtos/create-user.dto';

@Injectable()
export class CreateUserUsecase {
  constructor(
    @Inject(IUserRepository) private readonly userRepo: IUserRepository,
  ) {}

  @Transactional()
  async execute(dto: CreateUserRequestDto): Promise<CreateUserResponseDto> {
    const user = await this.userRepo.create(dto.toEntity());
    return { id: user.id! };
  }
}
